src/NHibernate.Test/Linq/ExpressionSessionLeakTest.cs

```diff
@@ -33,5 +34,36 @@ private WeakReference DoLinqInSeparateSession()
 				return new WeakReference(leakTest, false);
 			}
 		}
+
+		private static readonly PropertyInfo _sessionProperty = typeof(DefaultQueryProvider).GetProperty("Session", BindingFlags.NonPublic | BindingFlags.Instance);
+
+		[Test]
+		public void SessionIsNotResurrected()
```

Comment:
This test does not proof anything. It does not fail, with the old code.

Reply:
To make this test valid you need to remove `using` around session variable. Why? When you call `Dispose` the session is marked with suppress finilizer flag, so there is no "long" tracking then, no finilizer => nothing to await for.

Reply:
With the old code this was yielding `null` in case the session was disposed (and I had tested that), which would have always cause a null ref failure in normal use cases, because `Session` property is only called by code using it without checking for null. Due to the way this property is currently used, having `null` yielded should be considered a failure, because a null ref exception is in most case an almost meaningless error. (It is way better yielding the `InvalidOperationException`.)

Yes I was aware that resurrecting was requiring not disposing the session. But well, with the session tracking mechanism put in the test base class, we have a difficulty here. (Maybe time to solve it by the way, it could be related to troubles those weak-references tests causes with .Net Core.)